server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageStats.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.diskusage;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * The result of analyzing disk usage of each field in a shard/index
+ */
+public final class IndexDiskUsageStats implements ToXContentFragment, Writeable {
+    public static final String TOTAL = "total";
+    public static final String TOTAL_IN_BYTES = "total_in_bytes";
+    public static final String TERMS = "terms";
+    public static final String TERMS_IN_BYTES = "terms_in_bytes";
+    public static final String POSTINGS = "postings";
+    public static final String POSTINGS_IN_BYTES = "postings_in_bytes";
+    public static final String PROXIMITY = "proximity";
+    public static final String PROXIMITY_IN_BYTES = "proximity_in_bytes";
+    public static final String STORED_FIELDS = "stored_fields";
+    public static final String STORED_FIELDS_IN_BYTES = "stored_fields_in_bytes";
+    public static final String DOC_VALUES = "doc_values";
+    public static final String DOC_VALUES_IN_BYTES = "doc_values_in_bytes";
+    public static final String POINTS = "points";
+    public static final String POINTS_IN_BYTES = "points_in_bytes";
+    public static final String NORMS = "norms";
+    public static final String NORMS_IN_BYTES = "norms_in_bytes";
+    public static final String TERM_VECTORS = "term_vectors";
+    public static final String TERM_VECTORS_IN_BYTES = "term_vectors_in_bytes";
+
+    private final Map<String, PerFieldDiskUsage> fields;
+
+    public IndexDiskUsageStats() {
+        fields = new HashMap<>();
+    }
+
+    public IndexDiskUsageStats(StreamInput in) throws IOException {
+        this.fields = in.readMap(StreamInput::readString, PerFieldDiskUsage::new);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap(fields, StreamOutput::writeString, (o, v) -> v.writeTo(o));
+    }
+
+    PerFieldDiskUsage total() {
+        final PerFieldDiskUsage total = new PerFieldDiskUsage();
+        for (PerFieldDiskUsage value : fields.values()) {
+            total.add(value);
+        }
+        return total;
```

Comment:
I suspect that some users will be confused that this API returns a number of bytes that is slightly different from node/index stats since this API ignores some bytes that are read upon opening the index. Could we change the response so the total number of bytes reported is the actual size of the index (by summing up the lengths of files) and then have a `metadata` section for all bytes that are not attributed to a field by this API?

Reply:
Sure, I am on it.